### PR TITLE
Peter create todo list component

### PR DESCRIPTION
**Pull request by HeinzUBC**
Finished migrating TODOList module to using Redux Toolkit. Created rootReducer.js as a "combined" store area for the whole team. 

Most of the functionality for TODO module is complete. Users can view, edit, remove, and add the TODO items. Also, users can filter the TODO card list by what category (e.g. math, biology, work, etc) the cards belong to. 

image

Screen Shot 2023-06-04 at 10 22 00 PM

Screen Shot 2023-06-04 at 10 22 57 PM

However, currently, users can only filter cards if the cards have one of the "default" categories. In the near future, I will add functionality to enable filtering over user-created categories as well.

### CODE REVIEW

File: src/App.js
Comment (awctw) [inline src/App.js]: What happened to the SideBar?

File: src/Styles/TODOList.css
Comment (awctw) [inline src/Styles/TODOList.css:1]: We are using tailwind css so we don't need external css files as much, you should check it out, makes life easier. https://tailwindcss.com/docs/utility-first

File: src/store/rootReducer.js
Comment (awctw) [inline src/store/rootReducer.js]: This code already exists in main in the index.js file

File: src/components/TODOList/ListView.js
Comment (awctw) [inline src/components/TODOList/ListView.js:6]: you do not need to use the spread operator here to create a new array sortedTodos, just use `props.visibleTODOs` where needed

File: src/components/TODOList/TODOListMainView.js
Comment (awctw) [inline src/components/TODOList/TODOListMainView.js]: you do not need to set a variable here, just call `getVisibleTODOs` where needed

File: src/index.js
Comment (awctw) [inline src/index.js]: What linter are you using? We should all agree on one so that we do not get changes that are not really changes. Prettier is typically the recommended one for javascript

**Review by awctw — COMMENTED** [review]
There is a merge conflict which you see below.

File: src/App.js
Comment (HeinzUBC) [reply src/App.js]: I removed the siderbar component for purpose for testing my TODOList component in isolation. Once we merge this code in we can add it back :)

File: src/Styles/TODOList.css
Comment (HeinzUBC) [reply src/Styles/TODOList.css:1]: yes I understand. But I still need my TODOList.css file to style some of the elements. But I did use a lot of the pre-styled components from material tailwind CSS

File: src/store/rootReducer.js
Comment (HeinzUBC) [reply src/store/rootReducer.js]: Uh I changed it slightly. rootReducer.js now combines all the reducers with configureStore. In index.js, we now import the combined store (output of  configureStore) like this. Is this okay?

Screen Shot 2023-06-05 at 2 39 01 PM

File: src/components/TODOList/ListView.js
Comment (HeinzUBC) [reply src/components/TODOList/ListView.js:6]: Ok. I tried as you suggested. But because I am passing in the TODOList attribute of the TODOList redux slice, I cannot directly mutate it by calling sort() upon it. Hence I do have to create a local variable here.

File: src/components/TODOList/TODOListMainView.js
Comment (HeinzUBC) [reply src/components/TODOList/TODOListMainView.js]: Ok. Advice taken. Thanks!

File: src/index.js
Comment (HeinzUBC) [reply src/index.js]: I did not realize that my prettier plugin was not turned on. Anyhow, that is not fixed.

File: src/App.js
Comment (awctw) [reply src/App.js]: Please add it back in since you are trying to merge into main and make sure you component layout works with it.

File: src/Styles/TODOList.css
Comment (awctw) [reply src/Styles/TODOList.css:1]: I don’t mean the pre built components from material-tailwind but tailwindcss itself. It is a way of styling css, please check the link I sent above.

File: src/store/rootReducer.js
Comment (awctw) [reply src/store/rootReducer.js]: This is exactly what the current store does in the main branch. This is the index.js file in store folder and in the sec folder index.js we are doing the same: import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todo-slice";

const store = configureStore({
  reducer: { todoReducer },
});

export default store;

File: src/App.js
Comment (HeinzUBC) [reply src/App.js]: Ok. sure thing.

File: src/Styles/TODOList.css
Comment (HeinzUBC) [reply src/Styles/TODOList.css:1]: I saw it. I personally kinda of prefer the CSS sheet approach to that. Each html tag ends up having so much style info embedded in it. Also, this can be updated later on too. So this should not be a blocker?

**Review by awctw — COMMENTED** [review]
There is still a merge conflict, we cannot approve until that is resolved. Please run git pull and git merge and then resolve.

File: src/store/rootReducer.js
Comment (HeinzUBC) [reply src/store/rootReducer.js]: So this is really a question of which approach we should go with. I recommend this one because the code for combined the redux stores are all placed in rootReducer rather than index.js. It feelts better organized? What do you think?

File: src/Styles/TODOList.css
Comment (awctw) [reply src/Styles/TODOList.css:1]: It is not a blocker but our code style and UI will not be uniform if everyone else is using tailwindcss. Up to you when you would like to update, just letting you know. Your blocker is that merge conflict, nothing can be approved with it.

**Comment by HeinzUBC** [comment]
> There is still a merge conflict, we cannot approve until that is resolved. Please run git pull and git merge and then resolve.

File: src/Styles/TODOList.css
Comment (HeinzUBC) [reply src/Styles/TODOList.css:1]: Ok. I will migrate over to this style over time then

File: src/store/rootReducer.js
Comment (awctw) [reply src/store/rootReducer.js]: It is the same approach no?, only difference is the name of the file -> index.js vs rootReducer.js from what I see. I said to use the original in main purely because you will have to delete something in main already replacing it with the same logic.

File: src/App.js
Comment (HeinzUBC) [reply src/App.js]: Ok. Merge conflict has been resolved. Do you guys approve of it now? :)

File: src/store/rootReducer.js
Comment (HeinzUBC) [reply src/store/rootReducer.js]: But separating the store from index.js is better design right?

Comment (HeinzUBC) [reply src/store/rootReducer.js:1]: If you want to have the combined redux store be in index.js, I can make that change.

File: src/App.js
Comment (awctw) [inline src/App.js:1]: Where is your Todo component? It needs to be rendered on the same page as the SideBar.

File: src/index.js
Comment (awctw) [inline src/index.js:1]: Please see here: there was no actual change in the file, all that changed was that you imported Provider earlier in the list of imports. This is what I mean when the store and reducer setup is identical to what is in main already.

File: src/store/rootReducer.js
Comment (awctw) [reply src/store/rootReducer.js]: Please note there is a index.js in the store folder and a index.js in the src folder they are different files
This is your rootReducer.js:
```
import TODOListReducer from "./TODOListReducer";
import {configureStore} from "@reduxjs/toolkit";

export const store = configureStore({
    reducer: {
        todoReducer: TODOListReducer
    },
    devTools: true
});
```
This is what is inside the index.js file in the store folder:

```
import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todo-slice";

const store = configureStore({
  reducer: { todoReducer },
});

export default store;
```

No difference, the export is just following React customs for export and imports

File: src/store/TODOListReducer.js
Comment (awctw) [inline src/store/TODOListReducer.js:1]: The convention is to name the file with the word slice in it not reducer since the file contains a slice.

File: src/App.js
Comment (HeinzUBC) [reply src/App.js:1]: I removed it for now to remove the merge conflicts

File: src/store/rootReducer.js
Comment (HeinzUBC) [reply src/store/rootReducer.js]: Uhh... the index.js in the store has been removed in my branch

Comment (HeinzUBC) [reply src/store/rootReducer.js]: Can all of us just use the rootReducer to combine our various data slices?

File: src/App.js
Comment (awctw) [reply src/App.js:1]: You will need to have it rendered for progress-2 since they will be judging by the UI to make sure it works.

File: src/store/TODOListReducer.js
Comment (HeinzUBC) [reply src/store/TODOListReducer.js:1]: Advice taken

File: src/store/rootReducer.js
Comment (awctw) [reply src/store/rootReducer.js]: We can use it but its just weird that we are deleting code and then adding the exact same code back in with the only difference being index.js was renamed to rootReducer.

Comment (HeinzUBC) [reply src/store/rootReducer.js]: Well. Having two files both named index.js is a bit confusing haha

File: src/App.js
Comment (HeinzUBC) [reply src/App.js:1]: Yes. This merge is mainly to get most of the TODOList codebase into the repo so that others can also experiment with how best to position it in the dashboard view

**Review by awctw — APPROVED** [review]
Approved. Please migrate to TailwindCSS styling asap when free.

**Pull request merged**
